File: python/04/star_one.py
import os

input_file = os.path.join(os.path.dirname(__file__), "../../input/day_04.txt")

total = 0
with open(input_file, "r") as file:
    for x in file.readlines():
        x = x[:-1]
        win_nums, own_nums = x.split(":")[1].split("|")
        win_nums = set(win_nums.split(" "))
        win_nums.remove("")

        sub_count = 0.5
        for y in own_nums.split(" "):
            if y in win_nums:
                sub_count *= 2
        
        if sub_count != 0.5:
            total += int(sub_count)

print(total)

File: typescript/04/star_one.ts
import * as fs from "fs/promises";
import * as path from "path";

(async () => {
    const rawInput = await fs.readFile(path.resolve(__dirname, '../../input/day_04.txt'), "utf-8");
    const arr = rawInput.split("\n");

    const total = arr.reduce((sum, line) => {
        const [winNumsTmp, ownNums] = line.split(":")[1].split("|");
        const winNums = winNumsTmp.split(" ").filter(value => value != "")
        
        let sub_count = 0.5
        ownNums.split(" ").forEach(y => {
            if (winNums.includes(y))
                sub_count *= 2
        })

        return sum + ( sub_count != 0.5 ? sub_count : 0 )
    }, 0)
    console.log(total)
})();